### PR TITLE
Implementing memory

### PR DESCRIPTION
**Pull request by tischi**
Builds on https://github.com/haesleinhuepf/bia-tischi/pull/10

### CODE REVIEW

**Comment by tischi** [comment]
I think this could be merged now.

File: src/bia_tischi/_utilities.py [hunk 1]
```diff
@@ -57,16 +57,79 @@ def find_variable(variables, key, type_checker_function):
...
+    pricing += "- Last input: " + str(input_token) + " token = "
+    input_price = Models.usd_per_1k_input_token(model) * input_token / 10.0
+    pricing += print_costs_in_cent(input_price)
+    pricing += "- Last output: " + str(output_token) + " token = "
+    output_price = Models.usd_per_1k_output_token(model) * output_token / 10.0
+    pricing += print_costs_in_cent(output_price)
+    Context.session_price_us_cent += input_price + output_price
+    pricing += "- Total session costs: " + print_costs_in_cent(Context.session_price_us_cent)
+    pricing += "\n"
+    return pricing
```
Comment (haesleinhuepf) [inline src/bia_tischi/_utilities.py:121]: Out of curiosity: is this related to implementing _memory_? I would leave the pricing thing out for now as it might be a research topic itself.

File: setup.cfg [hunk 2]
```diff
@@ -39,6 +39,7 @@ install_requires =
     matplotlib
     scikit-learn
     tiktoken
+    langchain
```
Comment (haesleinhuepf) [inline setup.cfg:42]: I do not see code which uses langchain (except in a notebook). Is langchain really a dependency?

**Review by haesleinhuepf — COMMENTED** [review]
I'm struggling a bit to review this because changes related to memory are mixed with changes in other contexts (renaming functions, adding pricing, ...). I'm ok merging this. I'm just a big fan of concise PRs with only related changes. ;-)

File: src/bia_tischi/_utilities.py (hunk 1 above)
Comment (tischi) [reply src/bia_tischi/_utilities.py:121]: no it is not, I did bunch of stuff in parallel

File: setup.cfg (hunk 2 above)
Comment (tischi) [reply setup.cfg:42]: no, we can remove it

**Comment by tischi** [comment]
> I'm struggling a bit to review this because changes related to memory are mixed with changes in other contexts (renaming functions, adding pricing, ...). I'm ok merging this. I'm just a big fan of concise PRs with only related changes. ;-)

Yes, you are right, but it was so much that I changed that I did not see how to do this with individual PRs. I refactored almost everything....
